### PR TITLE
Simplify SendPacket API

### PR DESCRIPTION
**Pull request by AdityaSripal**
## Description

closes: #1395 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Review by damiannolan — APPROVED** [review]
Nice!! Looks super clean, love how much simpler the `SendTx` code in ICA is. 
Left one question about telemetry, not sure how critical it really is tho.

🚀

File: modules/apps/transfer/keeper/relay.go
Comment (chatton) [inline modules/apps/transfer/keeper/relay.go:106]: [q] where did these values end up previously? Is it safe to remove them?

File: modules/core/04-channel/keeper/packet.go
Comment (chatton) [inline modules/core/04-channel/keeper/packet.go:47]: [q] Could you help me understand why this check is a substitute for the previous `packet.GetDestPort() != channel.Counterparty.PortId` and `packet.GetDestChannel() != channel.Counterparty.ChannelId`? It's not immediately clear to me looking into `GetNextSequenceSend`

**Review by chatton — COMMENTED** [review]
Looks great! Nice cleanup. I just had a few small questions

**Review by crodriguezvega — COMMENTED** [review]
Thanks, @AdityaSripal! I think we need to add also some documentation in the migration docs for app developers, right? And we should decide in what version this should be released...

File: modules/core/04-channel/keeper/packet.go
Comment (AdityaSripal) [reply modules/core/04-channel/keeper/packet.go:47]: @crodriguezvega is correct. Yea the diffs on this PR look confusing in the merged view. But these checks have been removed completely since we are directly grabbing the destination info from the channelEnd

Before we were requiring callers to pass this in and checking it is the exact value we expect. But if we already have the value, then we could just plug it in ourselves. That is the point of this PR.

File: modules/apps/transfer/keeper/relay.go
Comment (AdityaSripal) [reply modules/apps/transfer/keeper/relay.go:85]: Yes. cc: @colin-axner

File: modules/core/04-channel/keeper/packet.go
Comment (chatton) [reply modules/core/04-channel/keeper/packet.go:47]: Ah I see, thanks for the clarification 👍

File: modules/core/04-channel/keeper/packet.go
Comment (colin-axner) [inline modules/core/04-channel/keeper/packet.go:30]: The sequence should be returned

File: modules/core/04-channel/keeper/packet.go
Comment (colin-axner) [inline modules/core/04-channel/keeper/packet.go:26]: I'm not sure how often we use `src`. Maybe we could just use `source`?

File: modules/core/04-channel/keeper/packet.go
Comment (asalzmann) [reply modules/core/04-channel/keeper/packet.go:30]: +1 please add this. we're currently doing this on a fork, but it would be easier for us to maintain if this was the default behavior https://github.com/Stride-Labs/ibc-go/pull/1/files

Comment (colin-axner) [reply modules/core/04-channel/keeper/packet.go:30]: Thanks for the feedback @asalzmann! I've noted it on #1969. I'll see if we can get someone to pick it up relatively soon

File: modules/apps/27-interchain-accounts/controller/keeper/relay.go
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/relay.go:42]: sequence should be returned by SendPacket so you can remove the get next sequence call

File: modules/apps/27-interchain-accounts/controller/keeper/relay.go
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/relay.go:27]: can you double check the test cases here, there might be some now that are unnecessary/illogical

File: modules/apps/27-interchain-accounts/types/expected_keepers.go
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/types/expected_keepers.go:24]: I think I'd prefer to just use `sourcePort` and `sourceChannel`

File: modules/apps/transfer/keeper/relay.go
Comment (colin-axner) [inline modules/apps/transfer/keeper/relay.go:80]: same here for checking unnecessary test cases

File: modules/apps/transfer/keeper/relay.go
Comment (colin-axner) [reply modules/apps/transfer/keeper/relay.go:85]: I think it is okay, so long as there is telemetry within core IBC showing where packets are being sent from/to, then the telemetry could filter on port ID to get the same info. This might break external UX though, so might make more sense to do in a followup

File: modules/core/04-channel/keeper/packet.go
Comment (colin-axner) [inline modules/core/04-channel/keeper/packet.go:113]: Hmm, I'd be careful here if we are also trying to return the sequence in the API, you don't want to make it so the nextSequenceSend is returned

Maybe it makes more sense to do:
```go
k.SetNextSequenceSend(ctx, srcPort, srcChannel, sequence+1)
```

Comment (colin-axner) [reply modules/core/04-channel/keeper/packet.go:113]: Technically you could be fine by ensuring you return packet.GetSequence(), but I think I'd actually prefer not to increment the sequence variable. 

You are using the variable for two different purposes:
- to get the sequence of the packet to send
- to set the next packet sequence to send

I'd rather not just the behaviour unexplicitly partway through the code

File: modules/core/05-port/types/module.go
Comment (colin-axner) [inline modules/core/05-port/types/module.go:115]: My preference is `sourcePort` `sourceChannel` since there isn't a reason not to be explicit

File: testing/endpoint.go
Comment (colin-axner) [reply testing/endpoint.go:378]: Can you open an issue to track this work?

File: modules/core/04-channel/keeper/packet.go
Comment (colin-axner) [reply modules/core/04-channel/keeper/packet.go:30]: When this change is made, a test should be added to ensure the correct sequence is returned (and not the next sequence send value)

File: modules/core/04-channel/keeper/packet.go
Comment (colin-axner) [reply modules/core/04-channel/keeper/packet.go:113]: Can you double check that we verify these values are correctly set in the tests?

**Review by colin-axner — COMMENTED** [review]
Concept ACK, see suggested feedback. Seems like a straightforward change. Nice improvement! :mechanical_arm:

**Comment by colin-axner** [comment]
Marking as draft just so it is easier to know when suggestions and merge conflicts have been addressed and the pr is ready for a full review again

File: modules/core/04-channel/keeper/packet.go
Comment (AdityaSripal) [reply modules/core/04-channel/keeper/packet.go]: Done. I ended up using packet.GetSequence() since that is what gets used in the rest of the lines, but I did switch to just setting sequence+1. So future changes don't affect anything.

I also added tests, so it would be caught anyway

File: modules/apps/transfer/keeper/relay.go
Comment (AdityaSripal) [reply modules/apps/transfer/keeper/relay.go:115]: I don't think telemetry is used on external UX? Also I just think it makes sense to keep it clean here, rather than grabbing channel just so that we can emit the same telemetry.

Afaiu, breaking telemetry does not count as an api breaking or state machine breaking change. So i don't want to be so restrictive and conservative in this case.

I'm in favor of just removing it. We can add it in the changelog so that people who need this information are informed of it beforehand

File: testing/endpoint.go
Comment (AdityaSripal) [reply testing/endpoint.go:430]: https://github.com/cosmos/ibc-go/issues/2452

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go
Comment (AdityaSripal) [inline modules/apps/27-interchain-accounts/controller/keeper/keeper.go:29]: No reason for us to duplicate ICS4Wrapper in the expected keepers. the interface already exists in port types. This also prevents accidentally missing a function. Done for all apps

File: modules/apps/27-interchain-accounts/types/expected_keepers.go
Comment (AdityaSripal) [inline modules/apps/27-interchain-accounts/types/expected_keepers.go:23]:
```suggestion
	GetChannel(ctx sdk.Context, srcPort, srcChan string) (channel channeltypes.Channel, found bool)
```

File: modules/apps/29-fee/types/expected_keepers.go
Comment (AdityaSripal) [inline modules/apps/29-fee/types/expected_keepers.go:19]:
```suggestion
	GetChannel(ctx sdk.Context, srcPort, srcChan string) (channel channeltypes.Channel, found bool)
```

File: modules/apps/transfer/types/expected_keepers.go
Comment (AdityaSripal) [inline modules/apps/transfer/types/expected_keepers.go:31]:
```suggestion
	GetChannel(ctx sdk.Context, srcPort, srcChan string) (channel channeltypes.Channel, found bool)
```

File: modules/core/04-channel/keeper/packet_test.go
Comment (AdityaSripal) [inline modules/core/04-channel/keeper/packet_test.go:23]: I had to change this back since it was conflicting with the local var in `TestSendPacket`. I also just find it makes more sense, since there are some non default values I put in the test cases.

**Review by crodriguezvega — COMMENTED** [review]
Thanks, @AdityaSripal!

We need to add these changes to the migration docs, but maybe that can be done in @damiannolan's [PR](https://github.com/cosmos/ibc-go/pull/2417).

We should also update all the documentation where `SendPacket` is mentioned; there are a few places...

**Review by colin-axner — CHANGES_REQUESTED** [review]
Needs migration docs. My main remaining concern is around telemetry. I wouldn't want to accidentally make a disruptive change, so we should ensure we have given sufficient consideration to the matter

File: docs/migrations/v5-to-v6.md
Comment (damiannolan) [inline docs/migrations/v5-to-v6.md:25]: I think you can remove this line

```suggestion
```

File: modules/core/04-channel/keeper/packet.go
Comment (damiannolan) [inline modules/core/04-channel/keeper/packet.go:22]: Should we update the godoc to mention that the sequence is now returned?

File: modules/core/05-port/types/module.go
Comment (damiannolan) [inline modules/core/05-port/types/module.go:119]: Should we leave these as unnamed return args as they are in other methods such as `GetAppVersion`?

**Review by damiannolan — APPROVED** [review]
LGTM! Nice work 🚀

File: modules/core/05-port/types/module.go
Comment (colin-axner) [reply modules/core/05-port/types/module.go:119]: It is good as is in my opinion. OnChanOpenInit has named returns. GetAppVersion probably should have had a named return

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/controller/ibc_middleware.go:236]: updated

**Comment by colin-axner** [comment]
@crodriguezvega Are there any changes left you think we should make? With regards to your current suggestions, I think it probably makes sense to continue using the packet variables so as to ensure we are referencing the correct values

Otherwise I think this pr is ready to be merged

File: docs/ibc/apps.md
Comment (crodriguezvega) [inline docs/ibc/apps.md:238]: I realized that this file is actually not used in the docs website, so could probably be removed...

File: modules/core/05-port/types/module.go
Comment (damiannolan) [reply modules/core/05-port/types/module.go:119]: Maybe we can make them all consistent in housekeeping followup.

**Comment by crodriguezvega** [comment]
> @crodriguezvega Are there any changes left you think we should make?

Yes, I just updated the documentation. I would appreciate a review to check it all looks ok. After that, I agree we can merge it.

File: docs/ibc/apps.md
Comment (colin-axner) [reply docs/ibc/apps.md:238]: I see, looks like we took this info and split it up into separate files? Lets open an issue whose task is to see if any info in here should be moved before deleting the file

Comment (colin-axner) [reply docs/ibc/apps.md:238]: #2480

**Review by colin-axner — APPROVED** [review]
Thanks @crodriguezvega for updating docs! I re-reviewed all changes

**Pull request merged**
